command/kv_list.go

```diff
@@ -75,25 +93,56 @@ func (c *KVListCommand) Run(args []string) int {
 		return 2
 	}
 
-	// Sanitize path
-	path := sanitizePath(args[0])
-	mountPath, v2, err := isKVv2(path, client)
-	if err != nil {
-		c.UI.Error(err.Error())
-		return 2
-	}
+	// If true, we're working with "-mount=secret foo" syntax.
+	// If false, we're using "secret/foo" syntax.
+	mountFlagSyntax := c.flagMount != ""
+
+	var (
+		mountPath   string
+		partialPath string
+		v2          bool
+	)
+
+	// Parse the paths and grab the KV version
+	if mountFlagSyntax {
+		// In this case, this arg is the secret path (e.g. "foo").
+		partialPath = sanitizePath(args[0])
+		mountPath, v2, err = isKVv2(sanitizePath(c.flagMount), client)
+		if err != nil {
+			c.UI.Error(err.Error())
+			return 2
+		}
 
-	if v2 {
-		path = addPrefixToKVPath(path, mountPath, "metadata")
+		if v2 {
+			partialPath = path.Join(mountPath, partialPath)
+		}
+	} else {
+		// In this case, this arg is a path-like combination of mountPath/secretPath.
+		// (e.g. "secret/foo")
+		partialPath = sanitizePath(args[0])
+		mountPath, v2, err = isKVv2(partialPath, client)
 		if err != nil {
 			c.UI.Error(err.Error())
 			return 2
 		}
 	}
 
-	secret, err := client.Logical().List(path)
+	// Add /data to v2 paths only
+	var fullPath string
+	if v2 {
+		fullPath = addPrefixToKVPath(partialPath, mountPath, "metadata")
+	} else {
+		// v1
+		if mountFlagSyntax {
+			fullPath = path.Join(mountPath, partialPath)
+		} else {
```

Comment:
It's a bit hard to follow these if blocks, perhaps they can be rearranged a bit for clarity? In particular we are doing `path.Join(...)` for kv-v1 here yet for kv-v2 in the previous `if` block.

Reply:
just copied the logic from [kv put](https://github.com/hashicorp/vault/blob/e4e9612fa42edc5e66c5c1528c65334da8c76062/command/kv_put.go#L154-L198)

each command in `kv ...` does this a little differently, going to look to consolidate the logic in a subsequent ticket